README
NAME
     Net::Netmask - parse, manipulate and lookup IP network blocks

SYNOPSIS
    use Net::Netmask;

     $block = new Net::Netmask (network block)
     $block = new Net::Netmask (network block, netmask)
     $block = new2 Net::Netmask (network block)
     $block = new2 Net::Netmask (network block, netmask)

     print $block;                  # a.b.c.d/bits
     print $block->base() 
     print $block->mask() 
     print $block->hostmask() 
     print $block->bits() 
     print $block->size() 
     print $block->maxblock()
     print $block->broadcast()
     print $block->next()
     print $block->match($ip);
     print $block->nth(1, [$bitstep]);

     if ($block->sameblock("network block")) ...
     if ($block->cmpblocks("network block")) ...

     $newblock = $block->nextblock([count]);

     for $ip ($block->enumerate([$bitstep])) { }

     for $zone ($block->inaddr()) { }

     my $table = {};
     $block->storeNetblock([$table])
     $block->deleteNetblock([$table])
     @missingblocks = $block->cidrs2inverse(@blocks)

     $block = findNetblock(ip, [$table])
     $block = findOuterNetblock(ip, [$table])
     @blocks = findAllNetblock(ip, [$table])
     if ($block->checkNetblock([$table]) ...
     $block2 = $block1->findOuterNetblock([$table])
     @blocks = dumpNetworkTable([$table])

     @blocks = range2cidrlist($beginip, $endip);
     @blocks = cidrs2cidrs(@blocks_with_dups)

     @listofblocks = cidrs2contiglists(@blocks);

     @blocks = sort @blocks
     @blocks = sort_network_blocks(@blocks)

     @sorted_ip_addrs = sort_by_ip_address(@unsorted_ip_addrs)

DESCRIPTION
    Net::Netmask parses and understands IPv4 CIDR blocks. It's built with an
    object-oriented interface. Nearly all functions are methods that operate
    on a Net::Netmask object.

    There are methods that provide the nearly all bits of information about
    a network block that you might want.

    There are also functions to put a network block into a table and then
    later lookup network blocks by IP address in that table. There are
    functions to turn a IP address range into a list of CIDR blocks. There
    are functions to turn a list of CIDR blocks into a list of IP addresses.

    There is a function for sorting by text IP address.

CONSTRUCTING
    Net::Netmask objects are created with an IP address and optionally a
    mask. There are many forms that are recognized:

    '216.240.32.0/24'               The preferred form.

    '216.240.32.0:255.255.255.0'
    '216.240.32.0-255.255.255.0'
    '216.240.32.0', '255.255.255.0'
    '216.240.32.0', '0xffffff00'
    '216.240.32.0 - 216.240.32.255'
    '216.240.32.4'                  A /32 block.

    '216.240.32'                    Always a /24 block.

    '216.240'                       Always a /16 block.

    '140'                           Always a /8 block.

    '216.240.32/24'
    '216.240/16'
    'default' or 'any'              0.0.0.0/0 (the default route)

    '216.240.32.0#0.0.31.255'       A hostmask (as used by Cisco
                                    access-lists).

    There are two constructor methods: "new" and "new2". The difference is
    that "new2" will return undef for invalid netmasks and "new" will return
    a netmask object even if the constructor could not figure out what the
    network block should be.

    With "new", the error string can be found as $block->{'ERROR'}. With
    "new2" the error can be found as Net::Netmask::errstr or
    $Net::Netmask::error.

METHODS
    ->desc()                 Returns a description of the network block. Eg:
                             216.240.32.0/19. This is also available as
                             overloaded stringification.

    ->base()                 Returns base address of the network block as a
                             string. Eg: 216.240.32.0. Base does not give an
                             indication of the size of the network block.

    ->mask()                 Returns the netmask as a string. Eg:
                             255.255.255.0.

    ->hostmask()             Returns the host mask which is the opposite of
                             the netmask. Eg: 0.0.0.255.

    ->bits()                 Returns the netmask as a number of bits in the
                             network portion of the address for this block.
                             Eg: 24.

    ->size()                 Returns the number of IP addresses in a block.
                             Eg: 256.

    ->broadcast()            The blocks broadcast address. (The last IP
                             address inside the block.) Eg: 192.168.1.0/24
                             => 192.168.1.255

    ->next()                 The first IP address following the block. (The
                             IP address following the broadcast address.)
                             Eg: 192.168.1.0/24 => 192.168.2.0

    ->first() & ->last()     Synonyms for ->base() and ->broadcast()

    ->match($ip)             Returns a true if the IP number $ip matches the
                             given network. That is, a true value is
                             returned if $ip is between base() amd
                             broadcast(). For example, if we have the
                             network 192.168.1.0/24, then

                               192.168.0.255 => 0
                               192.168.1.0   => "0 "
                               192.168.1.1   => 1
                               ...
                               192.168.1.255 => 255

                             $ip should be a dotted-quad (eg:
                             "192.168.66.3")

                             It just happens that the return value is the
                             position within the block. Since zero is a
                             legal position, the true string "0 " is
                             returned in it's place. "0 " is numerically
                             zero though. When wanting to know the position
                             inside the block, a good idiom is:

                               $pos = $block->match($ip) or die;
                               $pos += 0;

    ->maxblock()             Much of the time, it is not possible to
                             determine the size of a network block just from
                             it's base address. For example, with the
                             network block '216.240.32.0/27', if you only
                             had the '216.240.32.0' portion you wouldn't be
                             able to tell for certain the size of the block.
                             '216.240.32.0' could be anything from a '/23'
                             to a '/32'. The maxblock() method gives the
                             size of the largest block that the current
                             block's address would allow it to be. The size
                             is given in bits. Eg: 23.

    ->enumerate([$bitstep)   Returns a list of all the IP addresses in the
                             block. Be very careful not to use this function
                             of large blocks. The IP addresses are returned
                             as strings. Eg: '216.240.32.0', '216.240.32.1',
                             ... '216.240.32.255'.

                             If the optional argument is given, step through
                             the block in increments of a given network
                             size. To step by 4, use a bitstep of 30 (as in
                             a /30 network).

    ->nth($index, [$bitstep])
                             Returns the nth element of the array that
                             enumerate would return if it were called. So,
                             to get the first usable address in a block, use
                             nth(1). To get the broadcast address, use
                             nth(-1). To get the last usable adress, use
                             nth(-2).

    ->inaddr()               Returns an inline list of tuples. There is a
                             tuple for each DNS zone name in the block. If
                             the block is smaller than a /24, then the zone
                             of the enclosing /24 is returned.

                             Each tuple contains: the DNS zone name, the
                             last component of the first IP address in the
                             block in that zone, the last component of the
                             last IP address in the block in that zone.

                             Examples: the list returned for the block
                             '216.240.32.0/23' would be:
                             '32.240.216.in-addr.arpa', 0, 255,
                             '33.240.216.in-addr.arpa', 0, 255. The list
                             returned for the block '216.240.32.64/27' would
                             be: '32.240.216.in-addr.arpa', 64, 95.

    ->nextblock([$count])    Without a $count, return the next block of the
                             same size after the current one. With a count,
                             return the Nth block after the current one. A
                             count of -1 returns the previous block. Undef
                             will be returned if out of legal address space.

    ->sameblock($block)      Compares two blocks. The second block will be
                             auto-converted from a string if it isn't
                             already a Net::Netmask object. Returns 1 if
                             they are identical.

    ->cmpblocks($block)      Compares two blocks. The second block will be
                             auto-converted from a string if it isn't
                             already a Net::Netmask object. Returns -1, 0,
                             or 1 depending on which one has the lower base
                             address or which one is larger if they have the
                             same base address.

    ->contains($block)       Compares two blocks. The second block will be
                             auto-converted from a string if it isn't
                             already a Net::Netmask object. Returns 1 if the
                             second block fits inside the first block.
                             Returns 0 otherwise.

    ->storeNetblock([$t])    Adds the current block to an table of network
                             blocks. The table can be used to query which
                             network block a given IP address is in.

                             The optional argument allows there to be more
                             than one table. By default, an internal table
                             is used. If more than one table is needed, then
                             supply a reference to a HASH to store the data
                             in.

    ->deleteNetblock([$t])   Deletes the current block from a table of
                             network blocks.

                             The optional argument allows there to be more
                             than one table. By default, an internal table
                             is used. If more than one table is needed, then
                             supply a reference to a HASH to store the data
                             in.

    ->checkNetblock([$t])    Returns true of the netblock is already in the
                             network table.

    ->tag($name [, $value])  Tag network blocks with your own data. The
                             first argument is the name of your tag (hash
                             key) and the second argument (if present) is
                             the new value. The old value is returned.

    ->split($parts)          Splits a netmask into a number of sub
                             netblocks. This number must be a base 2 number
                             (2,4,8,16,etc.) and the number must not exceed
                             the number of IPs within this netmask.

                             e.g Net::Netmask->new( '10.0.0.0/24'
                             )->split(2) => ( Net::Netmask( '10.0.0.0/25') ,
                             Net::Netmask( '10.0.0.128/25' ) )

METHOD/FUNCTION COMBOS
    findOuterNetblock(ip, [$t])
                             Search the table of network blocks (created
                             with storeNetBlock) to find if any of them
                             contain the given IP address. The IP address
                             can either be a string or a Net::Netmask object
                             (method invocation). If more than one block in
                             the table contains the IP address or block, the
                             largest network block will be the one returned.

                             The return value is either a Net::Netmask
                             object or undef.

    cidrs2inverse(block, @listOfBlocks)
                             Given a block and a list of blocks,
                             cidrs2inverse() will return a list of blocks
                             representing the IP addresses that are in the
                             block but not in the list of blocks. It finds
                             the gaps.

                             The block will be auto-converted from a string
                             if it isn't already a Net::Netmask object. The
                             list of blocks should be Net::Netmask objects.

                             The return value is a list of Net::Netmask
                             objects.

OVERLOADING
    Overloading doesn't seem to work completeley on perl before version
    5.6.1. The test suite doesn't test overloading before that. At least for
    sort.

    ""                       Strinification is overloaded to be the ->desc()
                             method.

    cmp                      Numerical and string comparisions have been
                             overloaded to the ->cmpblocks() method. This
                             allows blocks to be sorted without specifying a
                             sort function.

FUNCTIONS
    sort_by_ip_address       This function is included in "Net::Netmask"
                             simply because there doesn't seem to be a
                             better place to put it on CPAN. It turns out
                             that there is one method for sorting
                             dotted-quads ("a.b.c.d") that is faster than
                             all the rest. This is that way. Use it as
                             "sort_by_ip_address(@list_of_ips)". That was
                             the theory anyway. Someone sent a faster
                             version ...

    sort_network_blocks      This function is a function to sort
                             Net::Netmask objects. It's faster than the
                             simpler "sort @blocks" that also works.

    findNetblock(ip, [$t])   Search the table of network blocks (created
                             with storeNetBlock) to find if any of them
                             contain the given IP address. The IP address is
                             expected to be a string. If more than one block
                             in the table contains the IP address, the
                             smallest network block will be the one
                             returned.

                             The return value is either a Net::Netmask
                             object or undef.

    findAllNetblock(ip, [$t])
                             Search the table of network blocks (created
                             with storeNetBlock) to find if any of them
                             contain the given IP address. The IP address is
                             expected to be a string. All network blocks in
                             the table that contain the IP address will be
                             returned.

                             The return value is a list of Net::Netmask
                             objects.

    dumpNetworkTable([$t])   Returns a list of the networks in a network
                             table (as created by ->storeNetblock()).

    range2cidrlist($startip, $endip)
                             Given a range of IP addresses, return a list of
                             blocks that span that range.


Changes

Revision history for Perl module Net::Netmask

1.9021 2014-07-17

 - Spelling fix in documentation.

1.9019 2013-10-01

 - Rename $b to $bits so that netmask can be sorted.  Change mostly
   from https://rt.cpan.org/Ticket/Display.html?id=76939

1.9018 2013-09-26

 - Minor MANIFEST fix.

1.9017 2013-09-20

 - Add network split() function.
 - Re-arrange the distribution a bit.

1.9016 2011-03-22

 - Fix bug #46996: warnings issued for bad input.
 - Fix bug #43348: use POSIX::floor() instead of int()
 - Rewrite netmask.t to use Test::More

1.9015 2006-11-30

 - Fix bug # 22662 reported by grjones at gmail: cidrs2inverse
   wouldn't notice /32-sized leftovers.

1.9014 2006-10-13

 - Fix bug # 22085 reported by grjones at gmail: cidrs2inverse
 - wouldn't notice /32-sized holes.

1.9013 2006-09-06

 - Added the nextblock() method as suggested by Robert Drake <rdrake at stayonline.net>

 - Bugfix: it couldn't parse 10/8 or 127/8

1.9011 2004-05-31

 - Some speed improvements from Todd R. Eigenschink <todd at tekinteractive.com>

1.9009 2004-04-12

 - Fix to netmasks.t for compatability with older perls

1.9008 2004-04-06

 - Added cidrs2inverse() which will find the gaps in a list of blocks.
 - Based on a request from Howard Jones <howard.jones at network-i.net>
   the tag() method was added.  It allows you to store your own data in
   a Net::Netmask object.  (Of course, you could have anyway as long as you
   didn't use the keys 'IBASE' or 'BITS')
 - Long ago, Alexandros M Manoussakis <amm2 at ini.cmu.edu> reported a bug
   that findAllNetblock would often return the same block multiple times.  
   Fixed.
 - Based on requests from Alexandros M Manoussakis <amm2 at ini.cmu.edu>
   and Lamprecht Andreas <andreas.a.lamprecht at siemens.com> the
   undefined behavior for overlapping blocks with cidrs2contiglists
   is no longer.  Such blocks will be in the same sublist.
 - Based on a requests from Tom Rudnick <trudnick at itc.nrcs.usda.gov>
   and Anthony Pardini <tony at pardini.org> new function was added:
   cidrs2cidrs().  cidrs2cidrs will condense a set of netblocks by
   combining blocks together that make up larger blocks.
 - Anthony Pardini <tony at pardini.org>, Frank Tegtmeyer <fte @ fte.to>
   and George Walker <gwalker at irish-times.com> pointed me to a bug
   with the contains() method.  Fixed.

1.9007 2004-01-01

 - At Max Baker <max at warped.org>'s request, the "require 5.6.1" 
   was removed for better compatability with older perl versions.

1.9006 2003-12-05

 - Removed '@'s from this file.

1.9005 2003-11-29

 - Matija Papec <matija at post.hinet.hr> suggested that I do a
   Schwartzian transform on the IP address sort function.  I 
   tried it.  It's faster.  Sort function replaced.

 - Added a sort_network_blocks function.
 - Added a contains() function to test if one block fits within
   another.
 - Peter Chen <petechen at netilla.com> was concerned about 
   using an illegal bitmask.  Now checked.
 - Long ago, Alexandros M Manoussakis <amm2 at ini.cmu.edu> noted that
   could be exported even though it was in EXPORT_OK.  Fixed.

1.9004 2003-05-28

 - Martin Lorensen <martin at lorensen.dk>: make 'any' a synonym for
   'default'.  
 - Bugfix (aslo from Martin): fix the require to accept 5.6.1

1.9003 2003-05-26

 - Roman Shishkin <roman at rt.mipt.ru> provided several (public exported)
   functions for looking at network tables:
   	dumpNetworkTable()
   	checkNetblock()
   Inspired: changing findOuterNetblock() so it can take a block
   as it's IP address.
 - Bugfix: notice that '218.0.0.0 - 221.255.255.255' isn't a valid
   netmask.  Reported by Dan Wright <wright at smx.pair.com>.
 - Bugfix: could not specify network '0.0.0.0-255.255.255.255'.  Fix
   from Dominic Mitchell <dom at semantico.com>.
 - Added ->sameblock() from Martin Lorensen <martin at lorensen.dk>
 - Added ->cmpblocks().
 - Added overloaded stringification so that blocks stringify to their
   description.
 - Added overloaded block comparision so that blocks can be compared and
   sorted.
 - Added hostmask syntax a.b.c.d#hostmask - Martin Lorensen <martin at lorensen.dk>
 - Bugfix: t/badnets.t was missing from the MANIFEST.
 - Some spelling and typo mistakes fixed in the documentation.

1.9002 2001-11-12

 - Change the license to make the Debian folks happy.  Interface through
   Jonas Smedegaard <dr at jones.dk>.

1.9001 2001-09-29

 - Sapient Fridge <sapient.fridge at WaCkY.zzn.com> and Alexander Karptsov
   <karp at visti.net> sent a patch for a bug in range2cidrlist.  The
   last IP in the range was skipped.
 - Sam Denton <sdenton at wantec.com> requested support for 
   a.b.c.d/mask.mask.mask.mask.
 - Sam also sent a request that I include the world's fastest
   sort-by-ip-address-in-perl function in Net::Netmask as there didn't
   seem to be a better place to put it.  I've included it.  The function
   in question was found/benchmarked by John Porter and written about
   in the Perl-Users Digest, Issue 3860, Volume 8.
 - Sam sent a patch to eliminate a couple of trailing spaces in
   the error codes.
 - My IP address are now 216.240.32/19 instead of 140.174.82/19 and
   thus I've changed the examples in the pod.  :-)

1.9 2001-05-15

 - Added deleteNetblock to match storeNetblock.
 - Carol Lerche <cafl at securify.com> contributed findOuterNetblock()
   and findAllNetblocks().
 - Kevin Baker <kevin at apnic.net> sent in patches that suggested a new
   handling of error conditions; extra error conditions to test for; and
   a test script to exercise the error conditions.
 - Bruce Peikon <Bruce.Peikon at msdw.com> sent a contribution which suggested
   that enumerate could do so by network.
 - Dominic Mitchell <dom at semantico.com> sent in code that suggested the 
   creation of cidrs2contiglists() and range2cidrlist().
 - A couple of documentation fixes from Igor Vinokurov <igor at rtsnet.ru>.

1.8 1999-09-20

 - Modified the match() method to return the position within the block.

1.7 1999-09-15


lib/Net/Netmask.pm
# Copyright (C) 1998-2006, David Muir Sharnoff <muir@idiom.org>

package Net::Netmask;

use vars qw($VERSION);
$VERSION = 1.9021;

require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(findNetblock findOuterNetblock findAllNetblock
	cidrs2contiglists range2cidrlist sort_by_ip_address
	dumpNetworkTable sort_network_blocks cidrs2cidrs
	cidrs2inverse);
@EXPORT_OK = (@EXPORT, qw(int2quad quad2int %quadmask2bits 
	%quadhostmask2bits imask sameblock cmpblocks contains));

my $remembered = {};
my %imask2bits;
my %size2bits;
my @imask;

# our %quadmask2bits;
# our %quadhostmask2bits;

use vars qw($error $debug %quadmask2bits %quadhostmask2bits);
$debug = 1;

use strict;
use warnings;
use Carp;
use POSIX qw(floor);
use overload
	'""' => \&desc,
	'<=>' => \&cmp_net_netmask_block,
	'cmp' => \&cmp_net_netmask_block,
	'fallback' => 1; 

sub new
{
	my ($package, $net, $mask) = @_;

	$mask = '' unless defined $mask;

	my $base;
	my $bits;
	my $ibase;
	undef $error;

	if ($net =~ m,^(\d+\.\d+\.\d+\.\d+)/(\d+)$,) {
		($base, $bits) = ($1, $2);
	} elsif ($net =~ m,^(\d+\.\d+\.\d+\.\d+)[:/](\d+\.\d+\.\d+\.\d+)$,) {
		$base = $1;
		my $quadmask = $2;
		if (exists $quadmask2bits{$quadmask}) {
			$bits = $quadmask2bits{$quadmask};
		} else {
			$error = "illegal netmask: $quadmask";
		}
	} elsif ($net =~ m,^(\d+\.\d+\.\d+\.\d+)[#](\d+\.\d+\.\d+\.\d+)$,) {
		$base = $1;
		my $hostmask = $2;
		if (exists $quadhostmask2bits{$hostmask}) {
			$bits = $quadhostmask2bits{$hostmask};
		} else {
			$error = "illegal hostmask: $hostmask";
		}
	} elsif (($net =~ m,^\d+\.\d+\.\d+\.\d+$,)
		&& ($mask =~ m,\d+\.\d+\.\d+\.\d+$,)) 
	{
		$base = $net;
		if (exists $quadmask2bits{$mask}) {
			$bits = $quadmask2bits{$mask};
		} else {
			$error = "illegal netmask: $mask";
		}
	} elsif (($net =~ m,^\d+\.\d+\.\d+\.\d+$,) &&
		($mask =~ m,0x[a-z0-9]+,i)) 
	{
		$base = $net;
		my $imask = hex($mask);
		if (exists $imask2bits{$imask}) {
			$bits = $imask2bits{$imask};
		} else {
			$error = "illegal netmask: $mask ($imask)";
		}
	} elsif ($net =~ /^\d+\.\d+\.\d+\.\d+$/ && ! $mask) {
		($base, $bits) = ($net, 32);
	} elsif ($net =~ /^\d+\.\d+\.\d+$/ && ! $mask) {
		($base, $bits) = ("$net.0", 24);
	} elsif ($net =~ /^\d+\.\d+$/ && ! $mask) {
		($base, $bits) = ("$net.0.0", 16);
	} elsif ($net =~ /^\d+$/ && ! $mask) {
		($base, $bits) = ("$net.0.0.0", 8);
	} elsif ($net =~ m,^(\d+\.\d+\.\d+)/(\d+)$,) {
		($base, $bits) = ("$1.0", $2);
	} elsif ($net =~ m,^(\d+\.\d+)/(\d+)$,) {
		($base, $bits) = ("$1.0.0", $2);
	} elsif ($net =~ m,^(\d+)/(\d+)$,) {
		($base, $bits) = ("$1.0.0.0", $2);
	} elsif ($net eq 'default' || $net eq 'any') {
		($base, $bits) = ("0.0.0.0", 0);
	} elsif ($net =~ m,^(\d+\.\d+\.\d+\.\d+)\s*-\s*(\d+\.\d+\.\d+\.\d+)$,) {
		# whois format
		$ibase = quad2int($1);
		my $end = quad2int($2);
		$error = "illegal dotted quad: $net" 
			unless defined($ibase) && defined($end);
		my $diff = ($end || 0) - ($ibase || 0) + 1;
		$bits = $size2bits{$diff};
		$error = "could not find exact fit for $net"
			if ! defined $error && (
				! defined $bits
				|| ($ibase & ~$imask[$bits]));
	} else {
		$error = "could not parse $net";
		$error .= " $mask" if $mask;
	}

	carp $error if $error && $debug;

	$ibase = quad2int($base || 0) unless defined $ibase;
	unless (defined($ibase) || defined($error)) {
		$error = "could not parse $net";
		$error .= " $mask" if $mask;
	}
	$ibase &= $imask[$bits]
		if defined $ibase && defined $bits;

	$bits = 0 unless $bits;
	if ($bits > 32) { 
		$error = "illegal number of bits: $bits"
			unless $error;
		$bits = 32;
	}

	return bless { 
		'IBASE' => $ibase,
		'BITS' => $bits, 
		( $error ? ( 'ERROR' => $error ) : () ),
	};
}

sub new2
{
	local($debug) = 0;
	my $net = new(@_);
	return undef if $error;
	return $net;
}

sub errstr { return $error; }
sub debug  { my $this = shift; return (@_ ? $debug = shift : $debug) }

sub base { my ($this) = @_; return int2quad($this->{'IBASE'}); }
sub bits { my ($this) = @_; return $this->{'BITS'}; }
sub size { my ($this) = @_; return 2**(32- $this->{'BITS'}); }
sub next { my ($this) = @_; int2quad($this->{'IBASE'} + $this->size()); }

sub broadcast 
{
	my($this) = @_;
	int2quad($this->{'IBASE'} + $this->size() - 1);
}

*first = \&base;
*last = \&broadcast;

sub desc 
{ 
	return int2quad($_[0]->{'IBASE'}).'/'.$_[0]->{'BITS'};
}

sub imask 
{
	return (2**32 -(2** (32- $_[0])));
}

sub mask 
{
	my ($this) = @_;

	return int2quad ( $imask[$this->{'BITS'}]);
}

sub hostmask
{
	my ($this) = @_;

	return int2quad ( ~ $imask[$this->{'BITS'}]);
}

sub nth
{
	my ($this, $index, $bitstep) = @_;
	my $size = $this->size();
	my $ibase = $this->{'IBASE'};
	$bitstep = 32 unless $bitstep;
	my $increment = 2**(32-$bitstep);
	$index *= $increment;
	$index += $size if $index < 0;
	return undef if $index < 0;
	return undef if $index >= $size;
	return int2quad($ibase+$index);
}

sub enumerate
{
	my ($this, $bitstep) = @_;
	$bitstep = 32 unless $bitstep;
	my $size = $this->size();
	my $increment = 2**(32-$bitstep);
	my @ary;
	my $ibase = $this->{'IBASE'};
	for (my $i = 0; $i < $size; $i += $increment) {
		push(@ary, int2quad($ibase+$i));
	}
	return @ary;
}

sub inaddr
{
	my ($this) = @_;
	my $ibase = $this->{'IBASE'};
	my $blocks = floor($this->size()/256);
	return (join('.',unpack('xC3', pack('V', $ibase))).".in-addr.arpa",
		$ibase%256, $ibase%256+$this->size()-1) if $blocks == 0;
	my @ary;
	for (my $i = 0; $i < $blocks; $i++) {
		push(@ary, join('.',unpack('xC3', pack('V', $ibase+$i*256)))
			.".in-addr.arpa", 0, 255);
	}
	return @ary;
}

sub tag
{
	my $this = shift;
	my $tag = shift;
	my $val = $this->{'T'.$tag};
	$this->{'T'.$tag} = $_[0] if @_;
	return $val;
}

sub quad2int
{
	my @bytes = split(/\./,$_[0]);

	return undef unless @bytes == 4 && ! grep {!(/\d+$/ && $_<256)} @bytes;

	return unpack("N",pack("C4",@bytes));
}

sub int2quad
{
	return join('.',unpack('C4', pack("N", $_[0])));
}

sub storeNetblock
{
	my ($this, $t) = @_;
	$t = $remembered unless $t;

	my $base = $this->{'IBASE'};

	$t->{$base} = [] unless exists $t->{$base};

	my $mb = maxblock($this);
	my $bits = $this->{'BITS'};
	my $i = $bits - $mb;

	$t->{$base}->[$i] = $this;
}

sub deleteNetblock
{
	my ($this, $t) = @_;
	$t = $remembered unless $t;

	my $base = $this->{'IBASE'};

	my $mb = maxblock($this);
	my $bits = $this->{'BITS'};
	my $i = $bits - $mb;

	return unless defined $t->{$base};

	undef $t->{$base}->[$i];

	for my $x (@{$t->{$base}}) {
		return if $x;
	}
	delete $t->{$base};
}

sub findNetblock
{
	my ($ipquad, $t) = @_;
	$t = $remembered unless $t;

	my $ip = quad2int($ipquad);
	return unless defined $ip;
	my %done;

	for (my $bits = 32; $bits >= 0; $bits--) {
		my $nb = $ip & $imask[$bits];
		next unless exists $t->{$nb};
		my $mb = imaxblock($nb, 32);
		next if $done{$mb}++;
		my $i = $bits - $mb;
		confess "$mb, $bits, $ipquad, $nb" if ($i < 0 or $i > 32);
		while ($i >= 0) {
			return $t->{$nb}->[$i]
				if defined $t->{$nb}->[$i];
			$i--;
		}
	}
	return undef;
}

sub findOuterNetblock
{
	my ($ipquad, $t) = @_;
	$t = $remembered unless $t;

	my $ip;
	my $mask;
	if (ref($ipquad)) {
		$ip = $ipquad->{IBASE};
		$mask = $ipquad->{BITS};
	} else {
		$ip = quad2int($ipquad);
		$mask = 32;
	}

	for (my $bits = 0; $bits <= $mask; $bits++) {
		my $nb = $ip & $imask[$bits];;
		next unless exists $t->{$nb};
		my $mb = imaxblock($nb, $mask);
		my $i = $bits - $mb;
		confess "$mb, $bits, $ipquad, $nb" if $i < 0;
		confess "$mb, $bits, $ipquad, $nb" if $i > 32;
		while ($i >= 0) {
			return $t->{$nb}->[$i]
				if defined $t->{$nb}->[$i];
			$i--;
		}
	}
	return undef;
}

sub findAllNetblock
{
	my ($ipquad, $t) = @_;
	$t = $remembered unless $t;
	my @ary ;
	my $ip = quad2int($ipquad);
	my %done;

	for (my $bits = 32; $bits >= 0; $bits--) {
		my $nb = $ip & $imask[$bits];
		next unless exists $t->{$nb};
		my $mb = imaxblock($nb, 32);
		next if $done{$mb}++;
		my $i = $bits - $mb;
		confess "$mb, $bits, $ipquad, $nb" if $i < 0;
		confess "$mb, $bits, $ipquad, $nb" if $i > 32;
		while ($i >= 0) {
			push(@ary,  $t->{$nb}->[$i])
				if defined $t->{$nb}->[$i];
			$i--;
		}
	}
	return @ary;
}

sub dumpNetworkTable
{
	my ($t) = @_;
	$t = $remembered unless $t;

	my @ary;
	foreach my $base (keys %$t) {
		push(@ary, grep (defined($_), @{$t->{base}}));
		for my $x (@{$t->{$base}}) {
			push(@ary, $x)
				if defined $x;
		}
	}
	return sort @ary;
}

sub checkNetblock
{
	my ($this, $t) = @_;
	$t = $remembered unless $t;

	my $base = $this->{'IBASE'};

	my $mb = maxblock($this);
	my $bits = $this->{'BITS'};
	my $i = $bits - $mb;

	return defined $t->{$base}->[$i];
}

sub match
{
	my ($this, $ip) = @_;
	my $i = quad2int($ip);
	my $imask = $imask[$this->{BITS}];
	if (($i & $imask) == $this->{IBASE}) {
		return (($i & ~ $imask) || "0 ");
	} else {
		return 0;
	}
}

sub maxblock 
{ 
	my ($this) = @_;
	return imaxblock($this->{'IBASE'}, $this->{'BITS'});
}

sub nextblock
{
        my ($this, $index) = @_;
	$index = 1 unless defined $index;
	my $newblock = bless {
		IBASE	=> $this->{IBASE} + $index * (2**(32- $this->{BITS})),
		BITS	=> $this->{BITS},
	};
	return undef if $newblock->{IBASE} >= 2**32;
	return undef if $newblock->{IBASE} < 0;
	return $newblock;
}

sub imaxblock
{
	my ($ibase, $tbit) = @_;
	confess unless defined $ibase;
	while ($tbit > 0) {
		my $im = $imask[$tbit-1];
		last if (($ibase & $im) != $ibase);
		$tbit--;
	}
	return $tbit;
}

sub range2cidrlist
{
	my ($startip, $endip) = @_;

	my $start = quad2int($startip);
	my $end = quad2int($endip);

	($start, $end) = ($end, $start)
		if $start > $end;
	return irange2cidrlist($start, $end);
}

sub irange2cidrlist
{
	my ($start, $end) = @_;
	my @result;
	while ($end >= $start) {
		my $maxsize = imaxblock($start, 32);
		my $maxdiff = 32 - floor(log($end - $start + 1)/log(2));
		$maxsize = $maxdiff if $maxsize < $maxdiff;
		push (@result, bless {
			'IBASE' => $start,
			'BITS' => $maxsize
		});
		$start += 2**(32-$maxsize);
	}
	return @result;
}

sub cidrs2contiglists
{
	my (@cidrs) = sort_network_blocks(@_);
	my @result;
	while (@cidrs) {
		my (@r) = shift(@cidrs);
		my $max = $r[0]->{IBASE} + $r[0]->size;
		while ($cidrs[0] && $cidrs[0]->{IBASE} <= $max) {
			my $nm = $cidrs[0]->{IBASE} + $cidrs[0]->size;
			$max = $nm if $nm > $max;
			push(@r, shift(@cidrs));
		}
		push(@result, [@r]);
	}
	return @result;
}

sub cidrs2cidrs
{
	my (@cidrs) = sort_network_blocks(@_);
	my @result;
	while (@cidrs) {
		my (@r) = shift(@cidrs);
		my $max = $r[0]->{IBASE} + $r[0]->size;
		while ($cidrs[0] && $cidrs[0]->{IBASE} <= $max) {
			my $nm = $cidrs[0]->{IBASE} + $cidrs[0]->size;
			$max = $nm if $nm > $max;
			push(@r, shift(@cidrs));
		}
		my $start = $r[0]->{IBASE};
		my $end = $max - 1;
		push(@result, irange2cidrlist($start, $end));
	}
	return @result;
}

sub cidrs2inverse
{
	my $outer = shift;
	$outer = __PACKAGE__->new2($outer) || croak($error) unless ref($outer);
	my (@cidrs) = cidrs2cidrs(@_);
	my $first = $outer->{IBASE};
	my $last = $first + $outer->size() -1;
	shift(@cidrs) while $cidrs[0] && $cidrs[0]->{IBASE} + $cidrs[0]->size < $first;
	my @r;
	while (@cidrs && $first <= $last) {
		if ($first < $cidrs[0]->{IBASE}) {
			if ($last <= $cidrs[0]->{IBASE}-1) {
				return (@r, irange2cidrlist($first, $last));
			}
			push(@r, irange2cidrlist($first, $cidrs[0]->{IBASE}-1));
		}
		last if $cidrs[0]->{IBASE} > $last;
		$first = $cidrs[0]->{IBASE} + $cidrs[0]->size;
		shift(@cidrs);
	}
	if ($first <= $last) {
		push(@r, irange2cidrlist($first, $last));
	}
	return @r;
}

sub by_net_netmask_block
{
	$a->{'IBASE'} <=> $b->{'IBASE'}
		|| $a->{'BITS'} <=> $b->{'BITS'};
}

sub sameblock
{
	return ! cmpblocks(@_);
}

sub cmpblocks
{
	my $this = shift;
	my $class = ref $this;
	my $other = (ref $_[0]) ? shift : $class->new(@_);
	return cmp_net_netmask_block($this, $other);
}

sub contains
{
	my $this = shift;
	my $class = ref $this;
	my $other = (ref $_[0]) ? shift : $class->new(@_);
	return 0 if $this->{IBASE} > $other->{IBASE};
	return 0 if $this->{BITS} > $other->{BITS};
	return 0 if $other->{IBASE} > $this->{IBASE} + $this->size -1;
	return 1;
}

sub cmp_net_netmask_block
{
	return ($_[0]->{IBASE} <=> $_[1]->{IBASE} 
		|| $_[0]->{BITS} <=> $_[1]->{BITS});
}

sub sort_network_blocks
{
	return
		map $_->[0],
		sort { $a->[1] <=> $b->[1] || $a->[2] <=> $b->[2] }
		map [ $_, $_->{IBASE}, $_->{BITS} ], @_;

}

sub sort_by_ip_address
{
	return
		map $_->[0],
		sort { $a->[1] cmp $b->[1] }
		map [ $_, pack("C4",split(/\./,$_)) ], @_;

}


sub split
{
	my ($self , $parts) = @_;

	my $num_ips = $self->size;

	confess "Parts must be defined and greater than 0."
		unless defined( $parts ) && $parts > 0;

	confess "Netmask only contains $num_ips IPs. Cannot split into $parts."
		unless $num_ips >= $parts; 
      
	my $log2 = log($parts) / log(2);
      
	confess "Parts count must be a number of base 2. Got: $parts"
		unless floor($log2) == $log2;

	my $new_mask = $self->bits + $log2;
      
	return 
		map { Net::Netmask->new( $_ . "/" .  $new_mask ) }
			map { $self->nth( ( $num_ips / $parts ) * $_ ) } 
				( 0 .. ( $parts - 1 ) );
}

BEGIN {
	for (my $i = 0; $i <= 32; $i++) {
		$imask[$i] = imask($i);
		$imask2bits{$imask[$i]} = $i;
		$quadmask2bits{int2quad($imask[$i])} = $i;
		$quadhostmask2bits{int2quad(~$imask[$i])} = $i;
		$size2bits{ 2**(32-$i) } = $i;
	}
}
1;
